### DISCUSSION TITLE
'hugo mod npm pack' breaks my customized 'package.json', do we need it?

### DISCUSSION THREAD

**Discussion by FuadEfendi** [razonyang/hugo-theme-bootstrap#747] [Q&A]
Hello @razonyang & Community,

I am trying to add new NPM modules to the project, but running `hugo mod npm pack` overrides `package.json` file and removes custom scripts.
These commands are mentioned in Theme installation guide:
```
hugo mod npm pack
npm install
```

I tried to fork HBS theme and created branch to have some additional features. Do I need to run `hugo mod npm pack' in a theme? In a hugo project? Unfortunately I cannot find much information, and it seems better to avoid this command...

Could you please provide more details? Thanks,

**Comment by FuadEfendi** [comment]
I found more details in "Hugo in Actions" book by Atishay Jain. Plus, some more guessing and experimentation. I think I got it.

If you want to add new dependency to Hugo module, add it to `package.hugo.json` file, and not to `package.json` (so that HBS theme currently breaks this rule... because `package.json` is not generated by `hugo mod npm pack`).

Hugo is modular system, and this is their way to generate cumulative `package.json` file for current project root directory from possibly many other `package.hugo.json` files in many other modules. 

Better to provide example: I need to use Algolia, and I added dependencies to `package.hugo.json` of my website project:
```
{
  "name": "MyCoolWebsite",
  "version": "0.1.0",
  "dependencies": {
    "@algolia/autocomplete-js": "1.7.1",
    "@algolia/autocomplete-plugin-query-suggestions": "1.7.1",
    "@algolia/autocomplete-plugin-recent-searches": "1.7.1",
    "@algolia/autocomplete-theme-classic": "1.7.1",
    "@algolia/client-search": "4.9.1"
  }
}
```

My website project depends on HBS theme:
```
theme:
  - github.com/razonyang/hugo-theme-bootstrap
```

I run command `hugo mod npm pack` and it generates bigger `package.json` file, which includes my custom Algolia imports, and HBS theme dependencies. Without this mechanism, if HBS theme adds new NPM module dependency, there is no way to have it automatically in our build settings `package.json`. So that, whenever you migrate from one version of theme to another, you need to run `hugo mod npm pack` and `npm install`, and commit `package.json` with `package-lock.json` to GitHub, so that Netlify build will use updated configs.

Documenting it here since it was really hard for me to find easy explanations of why we need this... thank you!

**Comment by razonyang** [comment]
Yes, we need to execute those commands on upgrading the theme as mentioned at https://hbs.razonyang.com/v1/en/docs/getting-started/upgradation/.

>  running hugo mod npm pack overrides package.json file and removes custom scripts.

`hugo mod npm pack` does override `package.json` since it merges the theme's `package.hugo.json` and your site's `package.hugo.json`, and then generate the `package.json` under your site root. You just need to define the dependencies and scripts inside your site's root `package.hugo.json`.

That part is missing from the documentation, I'll update the docs later.

EDITED: https://hbs.razonyang.com/v1/en/docs/getting-started/upgradation/#additional-dependencies
